### PR TITLE
erlang-r14: Incorporate patch from Erlbrew project

### PR DESCRIPTION
**Pull request by L2G**
This fixes build errors with llvm-gcc and clang.

### CODE REVIEW

**Comment by L2G** [comment]
The integration builds apparently failed because of a missing "sql.h" file. I have the unixodbc formula installed on my system, which includes sql.h, so that may be why it succeeded on mine but broke elsewhere.

I don't know whether it's more appropriate to patch Erlang to avoid a false positive detection of odbc or to simply declare a dependency on the unixodbc formula.

**Comment by L2G** [comment]
I'm now working on an additional patch to add a `--with-odbc` flag to the formula. If it is set, there will be a dependency on unixodbc, and a `--with-odbc` flag will be passed to configure. If it is _not_ set, then there will be an explicit `--without-odbc` flag passed to configure. Please let me know if a different approach would be preferred.

**Comment by DomT4** [comment]
@BrewTestBot test this please.

File: erlang-r14.rb [hunk 1]
```diff
@@ -10,11 +10,14 @@ class ErlangR14 < Formula
...
+  # Detection of odbc header files seems to be broken, so let the formula user
+  # decide whether or not this is needed.
+  option "with-odbc", "Also build the Erlang odbc application"
+
   depends_on "autoconf" => :build
   depends_on "automake" => :build
   depends_on "libtool" => :build
-
-  fails_with :llvm
+  depends_on "unixodbc" if build.with? "odbc"
```
Comment (DomT4) [inline erlang-r14.rb]: Can you also add a ` depends_on "openssl" ` here and ` revision 1 ` underneath the ` sha1 ` up top? Thanks!

File: erlang-r14.rb [hunk 2]
```diff
@@ -26,10 +29,15 @@ class ErlangR14 < Formula
     sha1 '86f76adee9bf953e5578d7998fda9e7dfc0d43f5'
   end
 
+  patch :p0, :DATA
```
Comment (DomT4) [inline erlang-r14.rb]: Can you add a note above this line explaining what the patch does, where it comes from, etc.

File: erlang-r14.rb [hunk 3]
```diff
@@ -53,6 +61,10 @@ def install
       args << "--enable-halfword-emulator" if build.include? 'halfword' # Does not work with HIPE yet. Added for testing only
     end
 
+    # Detection of odbc library and headers is slightly flaky, so be explicit about
+    # configuring it
+    args << "--#{build.with?("odbc") ? "with" : "without"}-odbc"
```
Comment (DomT4) [inline erlang-r14.rb]: ` args << (build.with?("odbc") ? "--with-odbc" : "--without-odbc") ` is preferred for easier reading.

File: erlang-r14.rb [hunk 4]
```diff
@@ -68,3 +80,50 @@ def install
     system "#{bin}/erl", "-noshell", "-eval", "crypto:start().", "-s", "init", "stop"
   end
 end
+
+# The patch following __END__ comes from the Erlbrew project (https://github.com/mrallen1/erlbrew)
```
Comment (DomT4) [inline erlang-r14.rb]: Snip this line.

**Comment by DomT4** [comment]
>  Please let me know if a different approach would be preferred.

We're trying to move away from ` if build.with? ` dependencies, as they're prone to being broken, but in this case I think it's relatively acceptable.

File: erlang-r14.rb [hunk 5]
```diff
@@ -10,11 +10,14 @@ class ErlangR14 < Formula
   option 'halfword', 'Enable halfword emulator (64-bit builds only)'
   option 'no-docs', 'Do not install documentation'
 
+  # Detection of odbc header files seems to be broken, so let the formula user
+  # decide whether or not this is needed.
+  option "with-odbc", "Also build the Erlang odbc application"
```
Comment (DomT4) [inline erlang-r14.rb]: Just ` option "with-odbc", "Build the Erlang odbc application" ` is fine :)

File: erlang-r14.rb (hunk 3 above)
Comment (L2G) [inline erlang-r14.rb]: Yep. I like that better.

**Comment by DomT4** [comment]
@mikemcquaid Any idea on this bot failure? We saw the same failure with the Erlang 16 PR earlier as well. It's an odd one:

```
==> Pouring erlang-r14-R14B04_1.mavericks.bottle.tar.gz
Error: /usr/local/Cellar/erlang/17.4.1 is not a directory
==> brew test --verbose Homebrew/versions/erlang-r14			 FAILED
Testing erlang-r14
==> /usr/local/Cellar/erlang-r14/R14B04_1/bin/erl -noshell -eval crypto:start(). -s init stop
/usr/local/Cellar/erlang-r14/R14B04_1/bin/erl: line 28: /private/tmp/erlang-r14-ytQc49/@@HOMEBREW_CELLAR@@/erlang-r14/R14B04_1/lib/erlang/erts-5.8.5/bin/erlexec: No such file or directory
/usr/local/Cellar/erlang-r14/R14B04_1/bin/erl: line 28: exec: /private/tmp/erlang-r14-ytQc49/@@HOMEBREW_CELLAR@@/erlang-r14/R14B04_1/lib/erlang/erts-5.8.5/bin/erlexec: cannot execute: No such file or directory
Error: erlang-r14: failed
```

**Comment by mikemcquaid** [comment]
@DomT4 Looks like it's storing an internal build path which is then deleted before the test is run.

**Comment by DomT4** [comment]
@mikemcquaid Is that something that needs poking upstream? 'Worst-case scenario' is we just don't bottle it for now, I guess, which isn't a departure from the current status quo on the Erlang formulae here but not a desirable thing. That way at least the paths stored would be the correct ones. Or perhaps we can inreplace the ` erl ` script with the expected ` prefix `, and try a bottle on that. I'll play with it a bit locally later this evening and see.

**Comment by mikemcquaid** [comment]
This doesn't seem to be a bottling issue, is it?

**Comment by mikemcquaid** [comment]
Can you see locally if the `test` passes? CC @jacknagel/@mistydemeo for `install_name_tool` thoughts.

**Comment by DomT4** [comment]
I presumed it might be, if the ` erl ` script is storing its install path in the binary itself and consequently making it a pain to bottle? It seems to be failing to execute the test because it expects to find certain paths still in ` /private/tmp `? Playing with the formula here is on my agenda for today; It slipped a bit from last night because I'm also frantically dashing to catch up on law work :smile_cat:

**Comment by DomT4** [comment]
Although I guess that's not bottle-specific, since it could be storing the buildpath outside of bottles as well, as we compile everything in ` /tmp ` first and then move it. Hmm. Will run it through a few builds soon and see.

**Comment by mikemcquaid** [comment]
Might be worth seeing if we ever had this version in core and how we handled (bottled?) it then.

**Comment by DomT4** [comment]
Homebrew/homebrew [shipped](https://github.com/Homebrew/homebrew/commit/881ec9bf18590e576e80d8b767158ab554cab783) ` R14B04 ` back in 2011, which is this version here.

At the time, you added [one bottle](https://github.com/Homebrew/homebrew/commit/67244a2bb3e65417943f6229f867c32e003c7ea2) to the formula the version before, which I presume was manually generated as it predated the bot, and that bottle was updated [here](https://github.com/Homebrew/homebrew/commit/580018b7f4bb2dd01720c5113ea832d599645f4d). The formula at the time doesn't seem to carry any special logic to handle the buildpath appearing in the binary issue, so perhaps the issue is new, somehow.

**Comment by DomT4** [comment]
From source build works fine, including the test.

```
🍺  /usr/local/Cellar/erlang-r14/R14B04_1: 6966 files, 260M, built in 34.2 minutes
brew test --verbose /Users/Dominyk/.brew_cache/Formula/erlang-r14.rb
Testing erlang-r14
==> /usr/local/Cellar/erlang-r14/R14B04_1/bin/erl -noshell -eval crypto:start(). -s init stop
```

The ` erl ` script the bot is struggling with reads:

```
ROOTDIR=/usr/local/Cellar/erlang-r14/R14B04_1/lib/erlang
BINDIR=$ROOTDIR/erts-5.8.5/bin
EMU=beam
PROGNAME=`echo $0 | sed 's/.*\///'`
export EMU
export ROOTDIR
export BINDIR
export PROGNAME
exec $BINDIR/erlexec ${1+"$@"}
```

Going to try building and installing a bottle next.

**Comment by DomT4** [comment]
Interesting:

```
brew install -v /Users/Dominyk/.brew_cache/Formula/erlang-r14.rb --build-bottle
brew bottle /Users/Dominyk/.brew_cache/Formula/erlang-r14.rb
brew uninstall erlang-r14
brew install /Users/Dominyk/Downloads/erlang-r14-R14B04_1.yosemite.bottle.tar.gz
==> Pouring erlang-r14-R14B04_1.yosemite.bottle.tar.gz
tar xf /Users/Dominyk/Downloads/erlang-r14-R14B04_1.yosemite.bottle.tar.gz
Error: /usr/local/Cellar/erlang/17.4.1 is not a directory
```

**Comment by DomT4** [comment]
@mikemcquaid It seems to be trying to load the Homebrew/homebrew ` erlang ` formulae when installing bottles for Homebrew/versions Erlang formulae?

```
 brew install -v ./erlang-r14-R14BO4_1.yosemite.bottle.tar.gz -d
/usr/local/Library/brew.rb (Formulary::BottleLoader): loading /usr/local/Library/Formula/erlang.rb
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/autoconf.rb
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/automake.rb
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/xz.rb
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/libtool.rb
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/openssl.rb
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/makedepend.rb
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/pkg-config.rb
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/unixodbc.rb
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/wxmac.rb
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/jpeg.rb
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/libpng.rb
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/libtiff.rb
==> Pouring erlang-r14-R14BO4_1.yosemite.bottle.tar.gz
tar xf /Users/Dominyk/Downloads/erlang-r14-R14BO4_1.yosemite.bottle.tar.gz
Error: /usr/local/Cellar/erlang/17.4.1 is not a directory
/usr/local/Library/Homebrew/keg.rb:99:in `initialize'
/usr/local/Library/Homebrew/formula_installer.rb:629:in `new'
/usr/local/Library/Homebrew/formula_installer.rb:629:in `pour'
/usr/local/Library/Homebrew/formula_installer.rb:167:in `install'
/usr/local/Library/Homebrew/cmd/install.rb:166:in `install_formula'
/usr/local/Library/Homebrew/cmd/install.rb:88:in `block in install'
/usr/local/Library/Homebrew/cmd/install.rb:88:in `each'
/usr/local/Library/Homebrew/cmd/install.rb:88:in `install'
/usr/local/Library/brew.rb:135:in `<main>'
```

**Comment by mikemcquaid** [comment]
Ah, course. In that case you'll need to fix the regex in `bottle_versions.rb` so it matches this bottle correctly.

**Comment by DomT4** [comment]
Is the bottle not covered by

```
    # e.g. x264-r2197.4.mavericks.bottle.tar.gz
    # e.g. lz4-r114.mavericks.bottle.tar.gz
    m = /-(r\d+\.?\d*)/.match(stem)
    return m.captures.first unless m.nil?
```

Or does that only permit a letter at the start of the second part, such as the ` R ` in ` R14BO4 ` but not the ` B `?

**Comment by mikemcquaid** [comment]
Evidently not, it seems. What you should probably do is add this specific filename as a test-case in `Library/Homebrew/test/test_bottle_versions.rb` and play around with `bottle_versions.rb` until all the tests cases pass (with `brew tests`).

**Comment by DomT4** [comment]
Not much luck here so far. Will keep working on it till I make some progress. If it takes longer than a few more days I may just merge this and generate bottles separately once I've ironed the issue out.

**Comment by mikemcquaid** [comment]
@DomT4 You got a branch somewhere I can look at/push to?

**Comment by DomT4** [comment]
Yeah, locally. I'll push it to the remote when I get near a computer again this evening.

**Comment by mikemcquaid** [comment]
@BrewTestBot test this please

**Comment by DomT4** [comment]
Merged! Thanks for your contribution to Homebrew, we appreciate it! :smiley_cat:

**Pull request closed without merging**